### PR TITLE
[SST-183] Fix V091 false positives in duplicate metric detection

### LINKED ISSUE #183: [Bug]: V091 duplicate metric detection ignores using_relationships, window, and non_additive_by
## Problem Description

V091 ("duplicate metric expression") compares only the `expr` field when detecting duplicate metrics. It ignores `using_relationships`, `window`, `non_additive_by`, and `derived`, causing false positives for metrics that share a base expression but are semantically different.

## Examples of False Positives

| Metric A | Metric B | Why different |
|----------|----------|---------------|
| `total_revenue` | `revenue_by_location` | Different `using_relationships` |
| `latest_order_total` | `max_order_value` | One has `non_additive_by` |
| `cumulative_revenue` | `customer_revenue_rank` | Different `window` configs |
| `total_orders` | `legacy_total_orders` | Different template syntax (`ref()` vs `column()`) |

## Expected Behavior

V091 should only flag metrics as duplicates when ALL of these match:
- `expr` (normalized)
- `using_relationships` (or both absent)
- `window` (or both absent)
- `non_additive_by` (or both absent)
- `derived` flag

## Actual Behavior

```
warning[SST-V091]: Metrics ['CUMULATIVE_REVENUE', 'CUSTOMER_REVENUE_RANK'] have identical expressions
```

These metrics have completely different window configurations but are flagged as duplicates.

## Environment

- SST version: 0.3.0
- Discovered in sst-jaffle-shop E2E testing

## Pre-submission Checklist

- [x] Searched existing issues to avoid duplicates
- [x] Included reproduction steps
- [x] Described expected vs actual behavior

### PR DESCRIPTION
**Pull request by mluizzi-whoop**
## Description

Fix V091 (duplicate metric detection) false positives. The validator now includes `using_relationships`, `window`, `non_additive_by`, and `derived` in the metric signature when comparing expressions. Metrics that share a base expression but differ in any of these modifiers are no longer flagged as duplicates.

## Related Issue

Closes #183

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

## Changes Made

- `duplicates.py`: Build a composite signature from `expr + using_relationships + window + non_additive_by + derived` instead of just `expr`
- Added 5 unit tests covering each modifier dimension

## Testing

- 1499 unit tests pass
- Verified against sst-jaffle-shop: false positives for `cumulative_revenue`/`customer_revenue_rank`, `revenue_by_location`/`total_revenue`, `latest_order_total`/`max_order_value` are eliminated

## PR Chain Position

PR #11 in chain. Merge order:
1. #167 → 2. #169 → 3. #170 → 4. #171 → 5. #173 → 6. #174 → 7. #175 → 8. #176 → 9. #181 → 10. #182 → 11. **This PR**

.... Generated with [Cortex Code](https://docs.snowflake.com/en/user-guide/cortex-code/cortex-code)

### CODE REVIEW

**Comment by mluizzi-whoop** [comment]
Addressed all 3 greptile findings in e2f9a27:
- **P1 non-canonical serialization**: Switched to `json.dumps(sort_keys=True)` for `window` and `non_additive_by` — produces stable key ordering
- **P2 pipe separator collision**: Changed delimiter from `|` to `\x00` (null byte) — cannot appear in SQL
- **P2 missing same-modifier test**: Added `test_same_expr_same_window_is_duplicate` — verifies identical modifiers still flag as duplicates

File: snowflake_semantic_tools/core/validation/rules/duplicates.py [hunk 1]
```diff
@@ -104,14 +104,22 @@ def _check_metric_duplicates(self, metrics_data: Dict, result: ValidationResult)
...
+                    if metric.get("using_relationships"):
+                        sig_parts.append(f"using:{sorted(metric['using_relationships'])}")
+                    if metric.get("window"):
+                        sig_parts.append(f"window:{metric['window']}")
+                    if metric.get("non_additive_by"):
+                        sig_parts.append(f"nab:{metric['non_additive_by']}")
+                    if metric.get("derived"):
+                        sig_parts.append("derived")
+                    signature = "|".join(sig_parts)
+                    expressions_seen[signature].append((i, name))
```
[the remark this answers is not in the document]
Comment (mluizzi-whoop) [reply snowflake_semantic_tools/core/validation/rules/duplicates.py:123]: Fixed in e2f9a27. Switched to `json.dumps(sort_keys=True)` for `window` and `non_additive_by` — produces deterministic key ordering regardless of insertion order.

[the remark this answers is not in the document]
Comment (mluizzi-whoop) [reply snowflake_semantic_tools/core/validation/rules/duplicates.py]: Fixed in e2f9a27. Changed delimiter from `|` to `\x00` (null byte) — cannot appear in SQL expressions, eliminating collision risk with Snowflake's `||` concatenation operator.

File: tests/unit/core/validation/test_duplicate_detection.py [hunk 2]
```diff
@@ -440,3 +440,59 @@ def test_multiple_duplicate_synonyms_all_reported(self, detector):
...
+        assert len(warnings) == 0
+
+    def test_same_expr_one_derived_not_duplicate(self, validator):
+        data = self._make_metrics_data([
+            {"name": "total", "expr": "SUM(orders.total)"},
+            {"name": "total_derived", "expr": "SUM(orders.total)", "derived": True},
+        ])
+        result = validator.validate(data)
+        warnings = [w for w in result.get_warnings() if w.rule_id == "SST-V091"]
+        assert len(warnings) == 0
```
[the remark this answers is not in the document]
Comment (mluizzi-whoop) [reply tests/unit/core/validation/test_duplicate_detection.py:508]: Fixed in e2f9a27. Added `test_same_expr_same_window_is_duplicate` — verifies two metrics with identical expr AND identical window config are still correctly flagged as V091 duplicates.

**Comment by mluizzi-whoop** [comment]
Superseded — all changes included in #167 (diagnostics overhaul chain)
